### PR TITLE
Reconnect and re-register computing node when session times out

### PR DESCRIPTION
**Pull request by soulasuna**
## Reconnect and re-register computing node when Proxy session times out.

Related to #16183.

Changes proposed in this pull request:
- Add session connection listener

### CODE REVIEW

File: shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-repository/shardingsphere-cluster-mode-repository-provider/shardingsphere-cluster-mode-repository-zookeeper-curator/src/main/java/org/apache/shardingsphere/mode/repository/cluster/zookeeper/listener/SessionConnectionListener.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    private static final int RECONNECT_INTERVAL_SECONDS = 5;
+    
+    private final InstanceDefinition instanceDefinition;
+    
+    private final ClusterPersistRepository repository;
+    
+    @Override
+    public void stateChanged(final CuratorFramework client, final ConnectionState connectionState) {
+        if (ConnectionState.LOST == connectionState) {
+            while (reRegister(client)) {
```
Comment (TeslaCN) [inline shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-repository/shardingsphere-cluster-mode-repository-provider/shardingsphere-cluster-mode-repository-zookeeper-curator/src/main/java/org/apache/shardingsphere/mode/repository/cluster/zookeeper/listener/SessionConnectionListener.java:47]: Why use `while` here?

Comment (soulasuna) [reply shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-repository/shardingsphere-cluster-mode-repository-provider/shardingsphere-cluster-mode-repository-zookeeper-curator/src/main/java/org/apache/shardingsphere/mode/repository/cluster/zookeeper/listener/SessionConnectionListener.java:47]: Here is an asynchronous process.
When session time out  will keep trying to connect to the registry will always try to connect to the registry every 5 seconds.
Return when re-registered.

Comment (TeslaCN) [reply shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-repository/shardingsphere-cluster-mode-repository-provider/shardingsphere-cluster-mode-repository-zookeeper-curator/src/main/java/org/apache/shardingsphere/mode/repository/cluster/zookeeper/listener/SessionConnectionListener.java:47]: But there is a `return` in the loop.

Comment (soulasuna) [reply shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-repository/shardingsphere-cluster-mode-repository-provider/shardingsphere-cluster-mode-repository-zookeeper-curator/src/main/java/org/apache/shardingsphere/mode/repository/cluster/zookeeper/listener/SessionConnectionListener.java]: Fixed

**Review by menghaoranss — APPROVED** [review]

**Pull request merged**